So the only error that mypy throws is this - 
`pandas/__init__.py:70: error: Module 'pandas._libs' has no attribute 'join'`

Theres a lot of other modules in those apis which do not particularly affect the reshape, stata PRs that i am working on but Ive I've included them in in case they need to be imported via top level from other modules in the future.

--- a/pandas/__init__.py
+++ b/pandas/__init__.py
@@ -45,30 +45,30 @@ from pandas._config import (get_option, set_option, reset_option,
 import pandas.core.config_init
 
 from pandas.core.api import (
-    #dtype
+    # dtype
     Int8Dtype, Int16Dtype, Int32Dtype, Int64Dtype, UInt8Dtype,
     UInt16Dtype, UInt32Dtype, UInt64Dtype, CategoricalDtype,
     PeriodDtype, IntervalDtype, DatetimeTZDtype,
 
-    #missing
+    # missing
     isna, isnull, notna, notnull,
 
-    #indexes
+    # indexes
     Index, CategoricalIndex, Int64Index, UInt64Index, RangeIndex,
     Float64Index, MultiIndex, IntervalIndex, TimedeltaIndex,
     DatetimeIndex, PeriodIndex, IndexSlice,
 
-    #tseries
+    # tseries
     NaT, Period, period_range, Timedelta, timedelta_range,
     Timestamp, date_range, bdate_range, Interval, interval_range,
     DateOffset,
 
-    #conversion
+    # conversion
     to_numeric, to_datetime, to_timedelta,
 
-    #misc
+    # misc
     np, TimeGrouper, Grouper, factorize, unique, value_counts,
-    array, Categorical, set_eng_float_format,  Series, DataFrame,
+    array, Categorical, set_eng_float_format, Series, DataFrame,
     Panel)
 
 from pandas.core.sparse.api import (
@@ -87,26 +87,26 @@ from pandas.core.reshape.api import (
 from pandas.util._print_versions import show_versions
 
 from pandas.io.api import (
-    #excel
+    # excel
     ExcelFile, ExcelWriter, read_excel,
 
-    #packers
+    # packers
     read_msgpack, to_msgpack,
 
-    #parsers
+    # parsers
     read_csv, read_fwf, read_table,
 
-    #pickle
+    # pickle
     read_pickle, to_pickle,
 
-    #pytables
+    # pytables
     HDFStore, read_hdf,
 
-    #sql
+    # sql
     read_sql, read_sql_query,
     read_sql_table,
 
-    #misc
+    # misc
     read_clipboard, read_parquet, read_feather, read_gbq,
     read_html, read_json, read_stata, read_sas)
 

--- a/pandas/__init__.py
+++ b/pandas/__init__.py
@@ -20,7 +20,9 @@ if missing_dependencies:
 del hard_dependencies, dependency, missing_dependencies
 
 # numpy compat
-from pandas.compat.numpy import *
+from pandas.compat.numpy import (
+    LooseVersion, tz_replacer, _tz_regex, np_datetime64_compat,
+    np_array_datetime64_compat, function)
 
 try:
     from pandas._libs import (hashtable as _hashtable,
@@ -43,17 +45,31 @@ from pandas._config import (get_option, set_option, reset_option,
 import pandas.core.config_init
 
 from pandas.core.api import (
+    #dtype
     Int8Dtype, Int16Dtype, Int32Dtype, Int64Dtype, UInt8Dtype,
-    UInt16Dtype, UInt32Dtype, UInt64Dtype, factorize, unique,
-    value_counts, isna, isnull, notna, notnull, CategoricalDtype,
-    PeriodDtype, IntervalDtype, DatetimeTZDtype, Categorical,
-    array, Grouper, set_eng_float_format, Index, CategoricalIndex,
-    Int64Index, UInt64Index, RangeIndex, Float64Index, MultiIndex,
-    IntervalIndex, TimedeltaIndex, DatetimeIndex, PeriodIndex,
+    UInt16Dtype, UInt32Dtype, UInt64Dtype, CategoricalDtype,
+    PeriodDtype, IntervalDtype, DatetimeTZDtype,
+
+    #missing
+    isna, isnull, notna, notnull,
+
+    #indexes
+    Index, CategoricalIndex, Int64Index, UInt64Index, RangeIndex,
+    Float64Index, MultiIndex, IntervalIndex, TimedeltaIndex,
+    DatetimeIndex, PeriodIndex, IndexSlice,
+
+    #tseries
     NaT, Period, period_range, Timedelta, timedelta_range,
     Timestamp, date_range, bdate_range, Interval, interval_range,
-    Series, DataFrame, Panel, IndexSlice, to_numeric, DateOffset,
-    to_datetime, to_timedelta, np, TimeGrouper)
+    DateOffset,
+
+    #conversion
+    to_numeric, to_datetime, to_timedelta,
+
+    #misc
+    np, TimeGrouper, Grouper, factorize, unique, value_counts,
+    array, Categorical, set_eng_float_format,  Series, DataFrame,
+    Panel)
 
 from pandas.core.sparse.api import (
     SparseArray, SparseDataFrame, SparseSeries, SparseDtype)
@@ -71,12 +87,28 @@ from pandas.core.reshape.api import (
 from pandas.util._print_versions import show_versions
 
 from pandas.io.api import (
-    read_clipboard, ExcelFile, ExcelWriter, read_excel,
-    read_feather, read_gbq, read_html, read_json,
-    read_msgpack, to_msgpack, read_parquet, read_csv,
-    read_fwf, read_table, read_pickle, to_pickle,
-    HDFStore, read_hdf, read_sas, read_sql, read_sql_query,
-    read_sql_table, read_stata)
+    #excel
+    ExcelFile, ExcelWriter, read_excel,
+
+    #packers
+    read_msgpack, to_msgpack,
+
+    #parsers
+    read_csv, read_fwf, read_table,
+
+    #pickle
+    read_pickle, to_pickle,
+
+    #pytables
+    HDFStore, read_hdf,
+
+    #sql
+    read_sql, read_sql_query,
+    read_sql_table,
+
+    #misc
+    read_clipboard, read_parquet, read_feather, read_gbq,
+    read_html, read_json, read_stata, read_sas)
 
 from pandas.util._tester import test
 import pandas.testing

--- a/pandas/__init__.py
+++ b/pandas/__init__.py
@@ -42,37 +42,42 @@ from pandas._config import (get_option, set_option, reset_option,
 # let init-time option registration happen
 import pandas.core.config_init
 
-
 from pandas.core.api import (
-    DataFrame, Series, array, Panel, Int64Index, Interval,
-    IntervalIndex, Float64Index, Index, DatetimeIndex, Int64Index,
-    MultiIndex, PeriodIndex, RangeIndex, TimedeltaIndex,
-    CategoricalIndex, Categorical, to_datetime, to_timedelta,
-    to_numeric, Timedelta, Timestamp, NaT, Period, isna,
-    CategoricalDtype, PeriodDtype, DatetimeTZDtype, IntervalDtype,
-    DateOffset, factorize, unique, value_counts, interval_range,
-    Grouper)
+    Int8Dtype, Int16Dtype, Int32Dtype, Int64Dtype, UInt8Dtype,
+    UInt16Dtype, UInt32Dtype, UInt64Dtype, factorize, unique,
+    value_counts, isna, isnull, notna, notnull, CategoricalDtype,
+    PeriodDtype, IntervalDtype, DatetimeTZDtype, Categorical,
+    array, Grouper, set_eng_float_format, Index, CategoricalIndex,
+    Int64Index, UInt64Index, RangeIndex, Float64Index, MultiIndex,
+    IntervalIndex, TimedeltaIndex, DatetimeIndex, PeriodIndex,
+    NaT, Period, period_range, Timedelta, timedelta_range,
+    Timestamp, date_range, bdate_range, Interval, interval_range,
+    Series, DataFrame, Panel, IndexSlice, to_numeric, DateOffset,
+    to_datetime, to_timedelta, np, TimeGrouper)
+
+from pandas.core.sparse.api import (
+    SparseArray, SparseDataFrame, SparseSeries, SparseDtype)
 
 from pandas.tseries.api import infer_freq
 from pandas.tseries import offsets
+
+from pandas.core.computation.api import eval
 
 from pandas.core.reshape.api import (
     concat, lreshape, melt, wide_to_long, merge, merge_asof,
     merge_ordered, crosstab, pivot, pivot_table, get_dummies,
     cut, qcut)
 
-from pandas.core.computation.api import eval
-
-from pandas.core.sparse.api import (
-    SparseArray, SparseDataFrame, SparseSeries, SparseDtype)
-
-from pandas.core.internals import BlockManager, _safe_reshape, make_block
-from pandas.core.arrays import DatetimeArray, IntervalArray, PeriodArray
-from pandas.core.arrays.sparse import BlockIndex, IntIndex
-
-from pandas._libs import join
-
 from pandas.util._print_versions import show_versions
+
+from pandas.io.api import (
+    read_clipboard, ExcelFile, ExcelWriter, read_excel,
+    read_feather, read_gbq, read_html, read_json,
+    read_msgpack, to_msgpack, read_parquet, read_csv,
+    read_fwf, read_table, read_pickle, to_pickle,
+    HDFStore, read_hdf, read_sas, read_sql, read_sql_query,
+    read_sql_table, read_stata)
+
 from pandas.util._tester import test
 import pandas.testing
 import pandas.arrays

